contracts/MechMarketplace.sol

```diff
@@ -489,7 +506,161 @@ contract MechMarketplace is IErrorsMarketplace {
+                revert HashNotApproved(recRequester, requestHash, signature);
+            }
+
+            return;
+        }
+
+        // Check EIP-1271 signature validity if requester is a contract
+        if (requester.code.length > 0) {
+            if (ISignatureValidator(requester).isValidSignature(requestHash, signature) == MAGIC_VALUE) {
+                return;
+            } else {
+                revert HashNotValidated(requester, requestHash, signature);
+            }
+        }
+
+        // For the correct ecrecover() function execution, the v value must be set to {0,1} + 27
+        // Although v in a very rare case can be equal to {2,3} (with a probability of 3.73e-37%)
+        // If v is set to just 0 or 1 when signing by the EOA, it is most likely signed by the ledger and must be adjusted
+        if (v < 4) {
+            // In case of a non-contract, adjust v to follow the standard ecrecover case
+            v += 27;
+        }
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            r := mload(add(signature, 32))
+            s := mload(add(signature, 64))
+        }
+
+        // Case of ecrecover with the request hash for EOA signatures
+        recRequester = ecrecover(requestHash, v, r, s);
+
+        // Final check is for the requester address itself
+        if (recRequester != requester) {
+            revert WrongRequesterAddress(recRequester, requester);
+        }
+    }
+
+    /// @dev Approves request hash for requester address.
+    /// @param hash Provided request hash to approve.
+    function requesterApproveHash(bytes32 hash) external {
+        mapRequesterApprovedHashes[msg.sender][hash] = true;
+        emit RequesterHashApproved(msg.sender, hash);
+    }
+    
+    /// @dev Delivers signed requests.
+    /// @notice This function must be called by mech delivering requests.
+    /// @param requester Requester address.
+    /// @param requestDatas Corresponding set of self-descriptive opaque request data-blobs.
+    /// @param signatures Corresponding set of signatures.
+    /// @param deliveryRates Corresponding set of actual charged delivery rates for each request.
+    /// @param deliveryDatas Corresponding set of self-descriptive opaque delivery data-blobs.
+    /// @param paymentData Additional payment-related request data, if applicable.
+    function deliverMarketplaceBatchWithSignatures(
+        address requester,
+        bytes[] memory requestDatas,
+        bytes[] memory signatures,
+        bytes[] memory deliveryDatas,
+        uint256[] memory deliveryRates,
+        bytes memory paymentData
+    ) external payable {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+        
+        // Check mech
+        address mechServiceMultisig = checkMech(msg.sender);
+
+        // TODO array length checks
+
+        // TODO check requester?
```

Comment:
Can requester be a service multisig as well?

Reply:
Ideally yes, we should see how this is handled in Safe